### ISSUE TITLE
Deadlock can occur when using MySQL backend.

### ISSUE THREAD

**Issue by ytsmiling** [optuna/optuna#1499]
## Expected behavior

Optuna handles MySQL errors.

## Environment

- Optuna version: 1.5.0
- Python version: 3.7
- OS: mac OS Catalina
- (Optional) Other libraries and their versions: MySQL + pymysql

## Error messages, stack traces, or logs

```
sqlalchemy.exc.InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(pymysql.err.OperationalError) (1213, 'Deadlock found when trying to get lock; try restarting transaction')
[SQL: INSERT INTO trial_user_attributes (trial_id, `key`, value_json) VALUES (%(trial_id)s, %(key)s, %(value_json)s)]
[parameters: {'trial_id': 2998, 'key': 'nodename', 'value_json': '"anoanymized"'}]
(Background on this error at: http://sqlalche.me/e/13/e3q8) (Background on this error at: http://sqlalche.me/e/13/7s2a)
```

## Steps to reproduce

1. Parallel optimization with MySQL backend.

## Reproducible examples (optional)

<details><summary>Code</summary>
<p>

```python
import os
import optuna
from multiprocessing import Pool


def f(x):
    return (x - 3) ** 2


# Objective function
def optuna_objective(trial):
    x = trial.suggest_uniform("x", -10, 10)
    y = f(x)

    nodename = os.uname().nodename
    trial.set_user_attr("nodename", nodename)

    print("x=%8.3f, y=%8.3f" % (x, y))

    return y


# Entry point
def main(args):
    rank, = args
    # Create a study
    ECHO=False
    storage = optuna.storages.RDBStorage(os.environ["OPTUNA_SQL"], engine_kwargs=dict(echo=ECHO))
    study = optuna.create_study(
        study_name="test_optuna_collision_01",
        storage=storage,
        load_if_exists=True,
        direction="minimize"
    )

    # Run optimization
    study.optimize(optuna_objective, n_trials=20)

    import time
    time.sleep(10)

    # Visualize the result
    if rank == 0:
        print("  #  %8s  %8s  %8s  %28s" % ("x", "y", "NODENAME", "STARTED"))
    for trial in study.trials:
        try:
            x = "%8.3f" % trial.params["x"]
            y = "%8.3f" % (trial.value or -1)
            if trial.value is None:
                print("NO_VALUE: ", trial.number)
            nodename = trial.user_attrs["nodename"]
            started = trial.datetime_start
        except (KeyError, AttributeError):
            x = ""
            y = ""
            nodename = ""
            started = ""
        if rank == 0:
            print("%3d: %s  %s  %8s  %28s" % (trial.number, x, y, nodename, started))


if __name__ == '__main__':
    with Pool(10) as pool:
        pool.map(main, [(i,) for i in range(10)])
```

</p>
</details>

## Additional context (optional)

See #1496 

I'll create a follow-up PR after #1490 and #1498 are merged.

**Comment by hvy** [comment]
@ytsmiling, is this issue resolved by https://github.com/optuna/optuna/pull/1514?

**Comment by ytsmiling** [comment]
I'm not sure whether we should close #1496, but I think this issue was solved by #1514, thanks.
